lib/bootsnap/cli/worker_pool.rb

```diff
@@ -0,0 +1,115 @@
+# frozen_string_literal: true
+
+module Bootsnap
+  class CLI
+    class WorkerPool
+      class << self
+        def create(size:, jobs:)
+          if size > 0 && Process.respond_to?(:fork)
+            new(size: size, jobs: jobs)
+          else
+            Inline.new(jobs: jobs)
+          end
+        end
+      end
+
+      class Inline
+        def initialize(jobs: {})
+          @jobs = jobs
+        end
+
+        def push(job, *args)
+          @jobs.fetch(job).call(*args)
+          nil
+        end
+
+        def spawn
+          # noop
+        end
+
+        def shutdown
+          # noop
+        end
+      end
+
+      class Worker
+        attr_reader :pipe, :pid
+
+        def initialize(jobs)
+          @jobs = jobs
+          @pipe_out, @pipe = IO.pipe
+          @pid = nil
+        end
+
+        def work_loop
+          loop do
+            job, *args = Marshal.load(@pipe_out)
+            return if job == :exit
+            @jobs.fetch(job).call(*args)
+          end
+        rescue IOError
+          nil
+        end
+
+        def spawn
+          @pid = Process.fork do
```

Comment:
Yes and no. There's a bit of IOs in the workload, so the GVL would be released a bit, but it's mostly a CPU heavy task, and the extensions don't release the GVL (because they instantiate Ruby objects).

So forking is the only way to get good performance. Maybe in the future Ractors could work.